components/dashboard/src/components/IdeOptions.tsx

```diff
@@ -0,0 +1,291 @@
+                        <span className="text-pk-content-primary capitalize">{ide.type}</span>
+                    </span>
+                </div>
+            ))}
+        </div>
+    );
+};
+
+export interface IdeOptionsModifyModalProps {
+    isLoading: boolean;
+    ideOptions: IDEOptions | undefined;
+    restrictedEditors: Set<string>;
+    pinnedEditorVersions: Map<string, string>;
+    updateMutation: UseMutationResult<
+        void,
+        Error,
+        { restrictedEditors: Set<string>; pinnedEditorVersions: Map<string, string> }
+    >;
+    onClose: () => void;
+}
+
+export const IdeOptionsModifyModal = ({
+    onClose,
+    updateMutation,
+    ideOptions,
+    ...props
+}: IdeOptionsModifyModalProps) => {
+    const orgLevelEditorVersionPinningEnabled = useFeatureFlag("org_level_editor_version_pinning_enabled");
+
+    const ideOptionsArr = useMemo(() => (ideOptions ? sortedIdeOptions(ideOptions) : undefined), [ideOptions]);
+    const pinnableIdes = useMemo(
+        () => ideOptionsArr?.filter((i) => orgLevelEditorVersionPinningEnabled && !!i.pinnable),
+        [ideOptionsArr, orgLevelEditorVersionPinningEnabled],
+    );
+
+    const [restrictedEditors, setEditors] = useState(props.restrictedEditors);
+    const [pinnedEditorVersions, setPinnedEditorVersions] = useState(props.pinnedEditorVersions);
+    const [editorVersions, setEditorVersions] = useState(new Map<string, string[]>());
+
+    useEffect(() => {
+        async function fetchData() {
+            if (!pinnableIdes) {
+                return;
+            }
+            const ideVersionsResult = await Promise.all(
+                pinnableIdes.map((ide) => getGitpodService().server.getIDEVersions(ide.id)),
+            );
+            const updatedVal = new Map<string, string[]>();
+            for (let i = 0; i < pinnableIdes.length; i++) {
+                const versions = ideVersionsResult[i]!;
+                updatedVal.set(pinnableIdes[i].id, versions);
+            }
+            setEditorVersions(updatedVal);
+        }
+        fetchData();
+    }, [pinnableIdes]);
+
+    const { toast } = useToast();
+
+    const handleUpdate = async () => {
+        if (computedError) {
+            return;
+        }
+        updateMutation.mutate(
+            { restrictedEditors, pinnedEditorVersions },
+            {
+                onSuccess: () => {
+                    toast({ message: "Editor options updated" });
+                    onClose();
+                },
+            },
+        );
+    };
+
+    const computedError = useMemo(() => {
+        if (!ideOptionsArr) {
+            return;
+        }
+        const leftOptions = ideOptionsArr.filter((i) => !restrictedEditors.has(i.id));
+        if (leftOptions.length === 0) {
+            return "At least one Editor has to be selected.";
+        }
+    }, [restrictedEditors, ideOptionsArr]);
+
+    const isLoading = props.isLoading || !pinnableIdes || pinnableIdes.length !== editorVersions.size;
```

Comment:
`pinnableIdes=false` will make it always show loading state, I'm going to push add a isLoading in fetchData if you're not around

Reply:
Update: false negative comment 🫣, it show loading forever maybe relates to some middle state of FF and React data. Refresh works